File: backend/src/main/java/com/daedan/festabook/place/dto/PlaceResponse.java
package com.daedan.festabook.place.dto;

import com.daedan.festabook.place.domain.Place;
import com.daedan.festabook.place.domain.PlaceAnnouncement;
import com.daedan.festabook.place.domain.PlaceCategory;
import com.daedan.festabook.place.domain.PlaceImage;
import com.daedan.festabook.timetag.domain.TimeTag;
import com.daedan.festabook.timetag.dto.TimeTagResponses;
import com.fasterxml.jackson.annotation.JsonFormat;
import java.time.LocalTime;
import java.util.List;

public record PlaceResponse(
        Long placeId,
        PlaceImageResponses placeImages,
        PlaceCategory category,
        String title,
        @JsonFormat(pattern = "HH:mm") LocalTime startTime,
        @JsonFormat(pattern = "HH:mm") LocalTime endTime,
        String location,
        String host,
        String description,
        PlaceAnnouncementResponses placeAnnouncements,
        TimeTagResponses timeTags
) {

    public static PlaceResponse from(
            Place place,
            List<PlaceImage> images,
            List<PlaceAnnouncement> announcements,
            List<TimeTag> timeTags
    ) {
        return new PlaceResponse(
                place.getId(),
                PlaceImageResponses.from(images),
                place.getCategory(),
                place.getTitle(),
                place.getStartTime(),
                place.getEndTime(),
                place.getLocation(),
                place.getHost(),
                place.getDescription(),
                PlaceAnnouncementResponses.from(announcements),
                TimeTagResponses.from(timeTags)
        );
    }

    public static PlaceResponse from(Place place) {
        return new PlaceResponse(
                place.getId(),
                PlaceImageResponses.from(List.of()),
                place.getCategory(),
                place.getTitle(),
                place.getStartTime(),
                place.getEndTime(),
                place.getLocation(),
                place.getHost(),
                place.getDescription(),
                PlaceAnnouncementResponses.from(List.of()),
                TimeTagResponses.from(List.of())
        );
    }
}

File: android/app/src/main/java/com/daedan/festabook/data/model/response/place/PlaceResponse.kt
package com.daedan.festabook.data.model.response.place

import com.daedan.festabook.data.model.response.place.PlaceGeographyResponse.PlaceCategory
import com.daedan.festabook.domain.model.Place
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class PlaceResponse(
    @SerialName("placeId")
    val id: Long,
    @SerialName("title")
    val title: String?,
    @SerialName("category")
    val category: PlaceCategory,
    @SerialName("description")
    val description: String?,
    @SerialName("imageUrl")
    val imageUrl: String?,
    @SerialName("location")
    val location: String?,
    @SerialName("timeTags")
    val timeTags: List<TimeTagResponse>,
)

fun PlaceResponse.toDomain() =
    Place(
        id = id,
        title = title,
        category = category.toDomain(),
        description = description,
        imageUrl = imageUrl,
        location = location,
        timeTags = timeTags.map { it.toDomain() },
    )
